java-checks-test-sources/src/main/java/checks/regex/InvalidRegexCheck.java

```diff
@@ -13,10 +13,26 @@ void noncompliant(String str) {
     str.replaceAll("x{1,2,3}|(", "x"); // Noncompliant [[sc=26;ec=27;secondary=13,13]] {{Fix the syntax errors inside this regex.}}
 
     str.matches("(\\w+-(\\d+)"); // Noncompliant [[sc=30;ec=31;secondary=15]] {{Fix the syntax error inside this regex.}}
+  }
+
+  void noncompliant_backreference(String str) {
+    str.matches("(\\w+)-\\2"); // Noncompliant {{Fix the back reference error inside this regex.}}
+
+    str.matches("(?<name>\\w+)-\\k<nae>"); // Noncompliant {{Fix the back reference error inside this regex.}}
 
-    str.matches("(\\w+)-\\2"); // False Negative - group numbers not handled in rule
+    str.matches(
+      "(?<g1>ab)"
+        + "\\k<g2>" // Noncompliant [[sc=12;ec=19;secondary=25,26]] {{Fix the back reference errors inside this regex.}}
+        + "\\k<g3>"
+        + "(?<g2>cd)"
+        + "(?<g3>ed)");
 
-    str.matches("(?<name>\\w+)-\\k<nae>"); // False Negative - group names not handled in rule
+    str.matches(
+      "(?<g1>ab)"
+        + "\\2" // Noncompliant [[sc=12;ec=15;secondary=32,33]] {{Fix the back reference errors inside this regex.}}
```

Comment:
Interestingly, unlike named back references, numbered back references don't cause an exception when they don't exist. They just don't match anything when the group doesn't exist yet. You could even produce a marginally useful regex that refers to a group defined later that can actually produce a match: `(?:(?:x|\\1)(.))+`. This will match strings where the first character is an x and every other odd-numbered character is either an x or equal to the previous character. The equivalent using named captures will throw an exception.

So we might consider either not complaining about numbered back references at all (so we'd only be complaining about errors that cause an exception) or only complaining if the numbered back reference refers to a group that does not exist in the regex at all (i.e. referring to a group defined later would be allowed when using numbers), in which case the regex would still not cause an exception, but definitely be wrong.

On the other hand, we might also just say "99% of the time referencing a not-yet-defined group is a mistake and the 1% FPs is acceptable" and just leave it as-is.

Reply:
On second thought, given that they do not cause an exception, it'd probably be best to not handle numeric back references here and instead handle them in the "this pattern can never match" rule.

Reply:
Fine by me, I'll only take into account named reference. And it will simplify the code of the rule as well!